pyomo/contrib/mindtpy/algorithm_base_class.py

```diff
@@ -2740,6 +2699,107 @@ def initialize_mip_problem(self):
         TransformationFactory('core.fix_integer_vars').apply_to(self.fixed_nlp)
         add_feas_slacks(self.fixed_nlp, config)
 
+    def initialize_subsolvers(self):
+        """Initialize and set options for MIP and NLP subsolvers."""
+        config = self.config
+        if config.mip_solver == 'gurobi_persistent' and config.single_tree:
+            self.mip_opt = GurobiPersistent4MindtPy()
+        else:
+            self.mip_opt = SolverFactory(config.mip_solver)
+        self.regularization_mip_opt = SolverFactory(config.mip_regularization_solver)
+        self.nlp_opt = SolverFactory(config.nlp_solver)
+        self.feasibility_nlp_opt = SolverFactory(config.nlp_solver)
+
+        self.check_subsolver_validity()
+        set_solver_mipgap(self.mip_opt, config.mip_solver, config)
+        set_solver_mipgap(
+            self.regularization_mip_opt, config.mip_regularization_solver, config
+        )
+
+        set_solver_constraint_violation_tolerance(
+            self.nlp_opt, config.nlp_solver, config
+        )
+        set_solver_constraint_violation_tolerance(
+            self.feasibility_nlp_opt, config.nlp_solver, config
+        )
+
+        self.set_appsi_solver_update_config()
+
+        if config.mip_solver == 'gurobi_persistent' and config.single_tree:
+            # PreCrush: Controls presolve reductions that affect user cuts
+            # You should consider setting this parameter to 1 if you are using callbacks to add your own cuts.
+            self.mip_opt.options['PreCrush'] = 1
+            self.mip_opt.options['LazyConstraints'] = 1
+
+        # set thread
+        if config.threads > 0:
+            self.mip_opt.options['threads'] = config.threads
+        if config.regularization_mip_threads > 0:
+            self.regularization_mip_opt['threads'] = config.threads
+
+        # regularization solver
+        if config.mip_regularization_solver == 'cplex':
+            if config.solution_limit is not None:
+                self.regularization_mip_opt.options[
+                    'mip limits solutions'
+                ] = config.solution_limit
+            self.regularization_mip_opt.options['mip strategy presolvenode'] = 3
+            # TODO: need to discuss if this option should be added.
```

Comment:
Let's discuss it now. Why do we want this option? If this is the one focused on obtaining good solutions, you can just reference our paper in Math Programming or the JOGO one here